Dynamizer/Lattice.hs

```diff
@@ -64,15 +81,57 @@ class Gradual p where
   static  :: p -> Sum Int
 
 instance Gradual (e (Ann a e)) => Gradual (Ann a e) where
-  lattice (Ann i e) = Ann i <$> lattice e
-  count   (Ann _ e) = count e
-  static  (Ann _ e) = static e
+  lattice (Ann i e)    = Ann i <$> lattice e
+  funLattice (Ann a e) = Ann a <$> funLattice e
+  count   (Ann _ e)    = count e
+  static  (Ann _ e)    = static e
 
-instance (Gradual t, Gradual e) => Gradual (ExpF t e) where
+instance (Gradual t, Gradual e, Dynamize t, Dynamize e) => Gradual (ExpF t e) where
   lattice = bitraverse lattice pure
   count   = bifoldMap count count
   static  = bifoldMap static static
 
+  funLattice (DConst name t e) = DConst name t <$> funLattice e
+  funLattice e@(DLam name args body t) =
+    DL.cons e $ DL.singleton $ DLam name args (dynamize body) $ dynamize t
```

Comment:
With the exception of the DLam case, all the other cases appear to to be equivalent to a traversable instance (might be the case we discussed). When I have dealt with cases like this, I would implement the special cases, then default to using traverse for everything else.

    funLattice e(@DLam name args body t) = ...
    funLattice e = traverse funLattice e

Reply:
done